director/auth/api_views.py

```diff
@@ -0,0 +1,144 @@
+        # Does basic validation before pinging Google to do token verification
+
+        exp = unverified_claims.get("exp")
+        if exp and float(exp) < time.time():
+            return bad_request("Token has expired")
+
+        if unverified_claims.get("iss") != GOOGLE_ISS:
+            return bad_request("Invalid token issuer")
+
+        if unverified_claims.get("aud") not in GOOGLE_AUDS:
+            return bad_request("Invalid token audience")
+
+        transport = requests.Request()
+        try:
+            claims = id_token.verify_token(token, transport)
+        except ValueError:
+            return bad_request("Token could not be verified")
+
+        if not claims.get("email_verified"):
+            return bad_request("Email address has not been verified")
+
+        email = claims.get("email")
+        given_name = claims.get("given_name")
+        family_name = claims.get("family_name")
+        name = claims.get("name")
+        if name is not None and given_name is None and family_name is None:
+            given_name, family_name = name.split()[:1]
+        try:
+            user = User.objects.get(email=email)
+        except ObjectDoesNotExist:
+            username = self.generate_username(email, given_name, family_name)
+            user = User.objects.create_user(
+                username, email=email, first_name=given_name, last_name=family_name
+            )
+
+        login(request, user, backend="django.contrib.auth.backends.ModelBackend")
+
+        token = jwt_encode_handler(jwt_payload_handler(user))
+        return Response({"token": token, "username": str(user)})
+
+    @staticmethod
+    def generate_username(
+        email: typing.Optional[str],
+        given_name: typing.Optional[str],
+        family_name: typing.Optional[str],
+    ) -> str:
+        """
+        Generate a username from an email address, given name and/or family name.
+
+        See tests for examples.
+        """
+
+        def check_name(name):
+            username = slugify(name)
+            if name and User.objects.filter(username=username).count() == 0:
+                return username
+            else:
+                return None
+
+        email_name = check_name(email.split("@")[0]) if email else None
+        if email_name:
+            return email_name
+
+        given_slug = check_name(given_name) if given_name else None
+        if given_slug:
+            return given_slug
+
+        name_slug = (
+            check_name(given_name + " " + family_name)
+            if given_name and family_name
+            else None
+        )
+        if name_slug:
+            return name_slug
+
+        email_slug = check_name(email) if email else None
+        if email_slug:
+            return email_slug
+
+        base_name = email_name if email_name else "user"
+        existing = User.objects.filter(username__startswith=base_name + "-").count()
```

Comment:
In theory this is OK, until we delete a user. That's why I suggested ordering by DESC and picking the highest one.